plugins/inputs/googlecoreiot/googlecoreiot.go

```diff
@@ -0,0 +1,436 @@
+		return err
+	}
+	h.listener = listener
+	h.Port = listener.Addr().(*net.TCPAddr).Port
+	switch h.Precision {
+
+	case "second":
+		h.handler.SetTimePrecision(time.Second)
+		break
+	case "microsecond":
+		h.handler.SetTimePrecision(time.Microsecond)
+		break
+	case "nanosecond":
+		h.handler.SetTimePrecision(time.Nanosecond)
+		break
+	default:
+		h.handler.SetTimePrecision(time.Millisecond)
+		break
+	}
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		server.Serve(h.listener)
+	}()
+
+	log.Printf("I! Started Google Core IoT service on %s\n", h.ServiceAddress)
+
+	return nil
+}
+
+/// Stop cleans up all resources
+func (h *GoogleListener) Stop() {
+	h.listener.Close()
+	h.wg.Wait()
+
+	log.Println("I! Stopped Google Core IoT service on ", h.ServiceAddress)
+}
+
+func (h *GoogleListener) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if req.URL.Path == h.Path {
+		h.AuthenticateIfSet(h.serveWrite, res, req)
+	} else {
+		h.AuthenticateIfSet(http.NotFound, res, req)
+	}
+}
+
+func (h *GoogleListener) decodeLineProtocol(payload []byte, obj Payload) ([]telegraf.Metric, error) {
+	var metrics []telegraf.Metric
+	var config = parsers.Config{}
+	config.DataFormat = h.DataFormat
+	parser, err := parsers.NewParser(&config)
+	if err != nil {
+		log.Println("E! Parser ", err)
+		return metrics, err
+
+	}
+	metrics, err = parser.Parse(payload)
+	if err != nil {
+		log.Println("E! Parser ", err)
+		return metrics, err
+	}
+
+	for key, value := range obj.Msg.Atts {
+		for _, m := range metrics {
+			m.AddTag(key, value.(string))
+			m.AddTag("message_id", obj.Msg.MessageID)
+			m.AddTag("message_id_2", obj.Msg.MessageID2)
+			m.AddTag("subscription", obj.Subscription)
+		}
+	}
+	return metrics, nil
+}
+
+func (h *GoogleListener) decodeJSON(payload []byte, obj Payload) (telegraf.Metric, error) {
+
+	e := JSONData{}
+	err := json.Unmarshal(payload, &e)
+	if err != nil {
+		log.Println("E! JSON Unmarshall ", err)
+		return h.handler.Metric()
+	}
+	b := []byte(strconv.FormatInt(e.Time, 10))
+	h.handler.SetTimestamp(b)
+	metrics, err := h.handler.Metric()
+	if err != nil {
+		log.Println("E! Metrics ", err)
+		return h.handler.Metric()
+	}
+	metrics.SetName(e.Name)
+	for key, value := range e.Fields {
+		metrics.AddField(key, value)
+	}
+
+	for key, value := range obj.Msg.Atts {
+
+		metrics.AddTag(key, value.(string))
+
+	}
+	metrics.AddTag("message_id", obj.Msg.MessageID)
+	metrics.AddTag("message_id_2", obj.Msg.MessageID2)
+	metrics.AddTag("subscription", obj.Subscription)
+	return metrics, nil
+
+}
+
+func (h *GoogleListener) serveNotFound(res http.ResponseWriter, req *http.Request) {
+	res.Header().Set("Content-Type", "application/json")
+	res.Header().Set("X-Influxdb-Version", "1.0")
+	res.WriteHeader(http.StatusNotFound)
+	res.Write([]byte(`{"error":"http: not found"}`))
+	return
+}
+
+func (h *GoogleListener) serveWrite(res http.ResponseWriter, req *http.Request) {
+	// Check that the content length is not too large for us to handle.
```

Comment:
This function will be called concurrently, you will need to add appropriate locking for shared data in the GoogleListener struct.